### PR TITLE
Restoring default includes  from configuration cache

### PR DESCRIPTION
**Pull request by 6hundreds**
Fixes https://github.com/gradle/gradle/issues/13665

### CODE REVIEW

**Comment by 6hundreds** [comment]
@bot-gradle test this

**Comment by 6hundreds** [comment]
@bot-gradle test this

**Comment by 6hundreds** [comment]
@bot-gradle cancel

**Comment by 6hundreds** [comment]
@bot-gradle test acc

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprint.kt [hunk 1]
```diff
@@ -78,7 +78,13 @@ sealed class ConfigurationCacheFingerprint {
         val value: Any?
     ) : ConfigurationCacheFingerprint()
 
-    data class RemoteScript(val uri: URI) : ConfigurationCacheFingerprint()
+    data class RemoteScript(
+        val uri: URI
+    ) : ConfigurationCacheFingerprint()
+
+    data class FileSystemDefaultExcludes(
```
Comment (6hundreds) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprint.kt:85]: DefaultExcludes naming sounds too generic for me. So decided to have some meaningful prefix

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt [hunk 2]
```diff
@@ -202,6 +203,9 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
                     return "remote script $uri has changed"
                 }
             }
+            is ConfigurationCacheFingerprint.FileSystemDefaultExcludes -> input.run {
+                host.updateFileSystemDefaultExcludes(excludes.toTypedArray())
```
Comment (6hundreds) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt:207]: DefaultExcludes can be changed in settings file. So if settings file was changed - configCache will got invalidated anyway. So I've considered it's barely possible to fail with checking was defaultExclude changed or not, and decided to update it at each check.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt [hunk 3]
```diff
@@ -725,6 +732,9 @@ class ConfigurationCacheFingerprintWriter(
         val capturedDirectories: MutableSet<File> = newConcurrentHashSet()
         val capturedFileSystemEntries: MutableSet<File> = newConcurrentHashSet()
 
+        private
+        val fileSystemDefaultExcludes = AtomicReference<Array<String>>()
```
Comment (6hundreds) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt:736]: As we having concurrent versions of collections for capturing, decided to have atomicReference for defaultExcludes

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileSystemDefaultExcludesTest.groovy [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import org.apache.tools.ant.DirectoryScanner
+import org.gradle.test.fixtures.file.TestFile
+
+class ConfigurationCacheFileSystemDefaultExcludesTest extends AbstractConfigurationCacheIntegrationTest {
```
Comment (mlopatkin) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileSystemDefaultExcludesTest.groovy:22]: Could you please add some tests with composite builds - changing excludes in the `buildSrc`, or a plugin, or a "normal" included build.

**Review by mlopatkin — CHANGES_REQUESTED** [review]
If I understand your approach correctly, you store the list of excludes when collecting the fingerprint and set them up when checking. There's no actual check though, as the user has to change settings to change the excludes, thus invalidating the cache. So we don't really need to fingerprint the excludes, it is just piggybacking.

I don't think this is a correct approach, as it introduces a side effect to the fingerprint checking, and overall looks like a violation of SRP. I suggest storing the excludes as part of the cached state (in the [ConfigurationCacheState](https://github.com/gradle/gradle/blob/master/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt) class), and restore them upon reading the state instead.

We can also try to replace the listeners with a pull-based API - ask some service what excludes we had after evaluating settings, and store the result in the cache.

One thing that isn't entirely clear to me, though, is how all of this works when composite builds and buildSrc are in play. I suspect that all builds have a chance to influence the resulting set of includes.

**Comment by 6hundreds** [comment]
> If I understand your approach correctly, you store the list of excludes when collecting the fingerprint and set them up when checking. There's no actual check though, as the user has to change settings to change the excludes, thus invalidating the cache. So we don't really need to fingerprint the excludes, it is just piggybacking.
> 
> I don't think this is a correct approach, as it introduces a side effect to the fingerprint checking, and overall looks like a violation of SRP. I suggest storing the excludes as part of the cached state (in the [ConfigurationCacheState](https://github.com/gradle/gradle/blob/master/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt) class), and restore them upon reading the state instead.
> 
> We can also try to replace the listeners with a pull-based API - ask some service what excludes we had after evaluating settings, and store the result in the cache.
> 
> One thing that isn't entirely clear to me, though, is how all of this works when composite builds and buildSrc are in play. I suspect that all builds have a chance to influence the resulting set of includes.

Yeah, I wouldn't wish to use fingerprints for this purpose, but this was a way that I'm familiar with. I will try another things.

Regarding last note - yes, this also a question for me and really each build can change includes, since includes are stored in memory of JVM (as Stefan told me).

File: subprojects/configuration-cache/build.gradle.kts [hunk 5]
```diff
@@ -79,6 +79,7 @@ dependencies {
     implementation(project(":build-option"))
 
     implementation(libs.asm)
+    implementation(libs.ant)
```
Comment (6hundreds) [inline subprojects/configuration-cache/build.gradle.kts:82]: This thing is bit concerning me, because I wouldn't add new dependency without need. Alternatively I could create wrapper/interface around DirectoryScanner, but this is a bit error prone IMO

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileSystemDefaultExcludesTest.groovy (hunk 4 above)
Comment (6hundreds) [reply subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileSystemDefaultExcludesTest.groovy:21]: @mlopatkin Is there a scenario we have to add here?

**Comment by 6hundreds** [comment]
@bot-gradle test acc

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileSystemDefaultExcludesTest.groovy (hunk 4 above)
Comment (mlopatkin) [reply subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileSystemDefaultExcludesTest.groovy:21]: Besides the obvious one with nothing fancy, I'd start with the included build that modifies the set of excludes. AFAIR, its settings file should be evaluated later than the root one. BuildSrc I'd test separately too, as its lifecycle is slightly different. Another corner case would be modifying the set of excludes inside the `build.gradle` - if you don't hit the check in the PatternSpecFactory, you can get the new state into the configuration cache at least.

It would also be nice to add some coverage for Ant tasks, or modify existing Ant test cases to examine the configuration cache. The latter should happen automatically to some extent with load-after-store, but given that the set of excludes is the static state that is not reliably reset after store - it makes sense to test multiple build runs too.

File: subprojects/configuration-cache/build.gradle.kts (hunk 5 above)
Comment (mlopatkin) [reply subprojects/configuration-cache/build.gradle.kts:82]: I don't think there's anything particularly bad with this dependency. Ant is part of the Gradle runtime regardless.

**Comment by mlopatkin** [comment]
To summarize our conversation:
1. To ensure the compatibility, we should capture the state of the excludes at the time the last settings file is evaluated, and not later; it is possible to modify the excludes later, though there are runtime checks in some places.
2. It seems to be natural to wrap the excludes state into a service that listens to `settingsEvaluated` and `afterStart` callbacks and pulls the state from the DirectoryScanner.
3. The clients (DefaultFileSystemFactory and PatternSpecFactory) receive updates by subscribing with some kind of `DefaultExcludesChangedListener`, the service from (2) broadcasts its updates
4. The configuration cache pulls the state from the service when saving and pushes upon restoring.

I'm not 100% sure how this interacts with updating the `DirectoryScanner` in `build.gradle` and calling Ant tasks afterward though, this should be tested. Then we can decide whether the observed behavior needs further fixes.

**Comment by 6hundreds** [comment]
@bot-gradle test acc

**Review by mlopatkin — CHANGES_REQUESTED** [review]
Thanks! I still have a rather major concern about the listener lifetime, and a few nitpicks.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt [hunk 6]
```diff
@@ -669,6 +678,13 @@ class ConfigurationCacheState(
         environmentChangeTracker.loadFrom(storedState)
     }
 
+    private
+    suspend fun DefaultWriteContext.writeFileSystemDefaultExcludes(gradle: GradleInternal) {
```
[the remark this answers is not in the document]
Comment (6hundreds) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:682]: Definitely

File: subprojects/core/src/main/java/org/gradle/internal/file/FileSystemDefaultExcludesProvider.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+
+/**
+ * Provides default excludes for file system traversing
+ */
+@ServiceScope(Scope.Global.class)
+public interface FileSystemDefaultExcludesProvider {
+
+    String[] getCurrentDefaultExcludes();
```
[the remark this answers is not in the document]
Comment (6hundreds) [reply subprojects/core/src/main/java/org/gradle/internal/file/FileSystemDefaultExcludesProvider.java:29]: I prefer List also, but keep it as array to reduce amount of transformations like array <-> list

[the remark this answers is not in the document]
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/internal/file/FileSystemDefaultExcludesProvider.java:29]: Well, we should convert array -> list down the line too, but this is only tangentially related.

**Comment by 6hundreds** [comment]
@bot-gradle test acc

**Review by mlopatkin — APPROVED** [review]
LGTM, thanks!

**Comment by 6hundreds** [comment]
@bot-gradle test and merge
